### DISCUSSION TITLE
tutorials/rendering/compositor

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#261] [User-contributed notes]
# tutorials/rendering/compositor

The compositor is a new feature in Godot 4 that allows control over the rendering pipeline when rendering the contents of a Viewport. It can be configured on a WorldEnvironment node where it applie...

https://docs.godotengine.org/en/4.3/tutorials/rendering/compositor.html

**Comment by SephReed** [comment]
Here's a link to a github discussion with various post-processing methods and info:
https://github.com/godotengine/godot/issues/99491

**Comment by mechalynx** [comment]
This is a pretty good listing of information on the Compositor I found, links to various resources and gives an overview for those interested in compositor effects and compute shaders: https://github.com/pink-arcana/godot-distance-field-outlines/discussions/1

Check the repo itself out too.

**Comment by hippogyz** [comment]
For C# users who meet ` @ void Godot.GodotObject.Dispose(bool): This function (free) can only be called from the render thread. ` error when dispose Compositor Effect, use `using` statement for `RenderSceneBuffersRD`. 
``` C#
    public override void _RenderCallback(int effectCallbackType, RenderData renderData)
    {
        // WRONG
        var renderSceneBuffers = renderData.GetRenderSceneBuffers() as RenderSceneBuffersRD;
        // ...
     }
```
``` C#
    public override void _RenderCallback(int effectCallbackType, RenderData renderData)
    {
        // Correct
        using var renderSceneBuffers = renderData.GetRenderSceneBuffers() as RenderSceneBuffersRD;
        // ...
     }
```

**Comment by OctagonIsAnAmazingShape** [comment]
Hi, this doesn't seem to work with the Mobile renderer on desktop (it complains about the color framebuffer not having TEXTURE_USAGE_STORAGE_BIT set (see [here](https://github.com/godotengine/godot/blob/876b290332ec6f2e6d173d08162a02aa7e6ca46d/servers/rendering/rendering_device.cpp#L3738)).

**Comment by Nodragem** [comment]
does `push_constant` work with DirectX 12? or is it Vulkan only?

**Reply by Calinou** [reply]
Push constants work in Vulkan, Direct3D 12 and Metal as they're abstracted away by RenderingDevice. In D3D12 terminology, the closest equivalent would be [root constants](https://learn.microsoft.com/en-us/windows/win32/direct3d12/using-constants-directly-in-the-root-signature) as I understand it.

Specialization constants also work in D3D12, even though D3D12 doesn't actually have a concept of specialization constants. See [this blog post](https://godotengine.org/article/d3d12-adventures-in-shaderland/#specialization-constants-creatively-working-around-the-lack-thereof-in-direct3d) for an explanation of how this was worked around in Godot.

**Reply by Nodragem** [reply]
Thank you for your reply! I am still a bit confused to when to use push constants as it seems that we can also declare uniform variables like `layout (location = 0) uniform float m;` in GLSL as explained this [introduction to compute shader](https://learnopengl.com/Guest-Articles/2022/Compute-Shaders/Introduction). Are push constants just more efficient?

**Reply by Calinou** [reply]
While they're accessed similarly in shaders, push constants are preferable to uniform buffers in some use cases: https://docs.vulkan.org/guide/latest/push_constants.html

Emphasis mine:

> A small bank of values writable via the API and accessible in shaders. Push constants allow the application to set values used in shaders **without creating buffers or modifying and binding descriptor sets for each update.**

**Comment by snail-with-tea** [comment]
Does Compositor work with Compatibility renderer? If so, what changes need to be done to shader?

**Reply by mechalynx** [reply]
According to the renderer overview page: https://docs.godotengine.org/en/latest/tutorials/rendering/renderers.html#environment-and-post-processing it isn't supported at the moment but support is planned: https://github.com/godotengine/godot-proposals/issues/7916#issuecomment-2932410245

**Comment by robostep9829** [comment]
Can this be used in mobile renderer for _anything_? Since the color texture there doesn't have the storage flag, it can only be written by fragment shaders, not compute. Also it doesn't have the "copy to" flag neither in forward+ nor mobile renderer.
